File: RaftKotlin/raft-node/src/main/kotlin/raft/transport/grpc/services/ClientService.kt
package raft.transport.grpc.services

import com.google.protobuf.Empty
import raft.Client
import raft.RaftClientGrpcKt
import raft.core.node.RaftNode
import raft.core.utils.client.*
import raft.diagnosticsResponse
import raft.putResponse
import raft.getResponse
import raft.serverStateResponse
import raft.serverTermResponse
import raft.transport.grpc.utils.extensions.toGRPC
import java.time.Instant

/**
 * gRPC service implementation for Raft client operations.
 */
class ClientService(
    private val node: RaftNode
) : RaftClientGrpcKt.RaftClientCoroutineImplBase() {

    override suspend fun put(request: Client.PutRequest): Client.PutResponse {
        val response = node.put(request = PutRequest(key = request.key, value = request.value))

        return putResponse {
            success = response.success
            response.leaderHint?.let { hint ->
                leaderHint = hint.toGRPC()
            }
        }
    }

    override suspend fun get(request: Client.GetRequest): Client.GetResponse {
        val response = node.get(request = GetRequest(key = request.key))

        return getResponse {
            response.value?.let { value = it }
            response.leaderHint?.let { hint ->
                leaderHint = hint.toGRPC()
            }
        }
    }

    override suspend fun getDebug(request: Client.GetRequest): Client.GetResponse {
        val response = node.getDebug(request = GetRequest(key = request.key))

        return getResponse {
            response.value?.let { value = it }
            response.leaderHint?.let { hint ->
                leaderHint = hint.toGRPC()
            }
        }
    }

    override suspend fun getServerState(request: Empty): Client.ServerStateResponse {
        val response = node.getState()

        return serverStateResponse {
            id = response.id
            state = response.state.toGRPC()
        }
    }

    override suspend fun getServerTerm(request: Empty): Client.ServerTermResponse {
        val response = node.getTerm()

        return serverTermResponse {
            term = response.term.toLong()
            id = response.id
        }
    }

    override suspend fun getDiagnostics(request: Client.DiagnosticsRequest): Client.DiagnosticsResponse {
        val response = node.getDiagnostics(request = DiagnosticsRequest(
            start = Instant.ofEpochSecond(request.startTime.seconds, request.startTime.nanos.toLong()),
            end = Instant.ofEpochSecond(request.endTime.seconds, request.endTime.nanos.toLong()))
        )

        return diagnosticsResponse {
            id = response.id
            implementation = response.implementation
            version = response.version
            compactionThreshold = response.compactionThreshold
            metrics.addAll(response.metrics.toGRPC())
        }
    }
}

File: RaftSwift/Sources/RaftGRPCTransport/Services/ClientService.swift
import Foundation
import GRPCCore
import RaftCore
import SwiftProtobuf

struct ClientService: Raft_RaftClient.SimpleServiceProtocol {
    let node: any RaftNodeProtocol

    init(node: any RaftNodeProtocol) {
        self.node = node
    }

    func put(request: Raft_PutRequest, context: ServerContext) async throws -> Raft_PutResponse {
        let response = try await node.put(request: PutRequest(key: request.key, value: request.value))

        return .with { grpcResponse in
            grpcResponse.success = response.success
            if let leaderHint = response.leaderHint {
                grpcResponse.leaderHint = leaderHint.toGRPC()
            }
        }
    }

    func get(request: Raft_GetRequest, context: ServerContext) async throws -> Raft_GetResponse {
        let response = await node.get(request: GetRequest(key: request.key))

        return .with { grpcResponse in
            if let value = response.value {
                grpcResponse.value = value
            }
            if let leaderHint = response.leaderHint {
                grpcResponse.leaderHint = leaderHint.toGRPC()
            }
        }
    }

    func getDebug(request: Raft_GetRequest, context: ServerContext) async throws -> Raft_GetResponse {
        let response = await node.getDebug(request: GetRequest(key: request.key))

        return .with { grpcResponse in
            if let value = response.value {
                grpcResponse.value = value
            }
            if let leaderHint = response.leaderHint {
                grpcResponse.leaderHint = leaderHint.toGRPC()
            }
        }
    }

    func getServerState(request: Google_Protobuf_Empty, context: ServerContext) async throws -> Raft_ServerStateResponse {
        let response = await node.getState()

        return .with { grpcResponse in
            grpcResponse.id = UInt32(response.id)
            grpcResponse.state = response.state.toGRPC()
        }
    }

    func getServerTerm(request: Google_Protobuf_Empty, context: ServerContext) async throws -> Raft_ServerTermResponse {
        let response = await node.getTerm()

        return .with { grpcResponse in
            grpcResponse.term = UInt64(response.term)
            grpcResponse.id = UInt32(response.id)
        }
    }

    func getDiagnostics(request: Raft_DiagnosticsRequest, context: ServerContext) async throws -> Raft_DiagnosticsResponse {
        let response: DiagnosticsResponse = await node.getDiagnostics(request: DiagnosticsRequest(
            start: request.startTime.date,
            end: request.endTime.date,
        ))

        return .with { grpcResponse in
            grpcResponse.id = UInt32(response.id)
            grpcResponse.implementation = response.implementation
            grpcResponse.version = response.version
            grpcResponse.compactionThreshold = UInt32(response.compactionThreshold)
            grpcResponse.metrics = response.metrics?.toGRPC() ?? []
        }
    }
}
